### ISSUE TITLE
Improve url_fuzzer plugin using tricks from bfac

### ISSUE THREAD

**Issue by andresriancho** [andresriancho/w3af#16271]
There is a nice tool called [bfac](https://github.com/mazen160/bfac) which has some url fuzzing tricks which are not implemented in [url_fuzzer](https://github.com/andresriancho/w3af/blob/a53d59b73c493013a90fccf468cf443e47f869d0/w3af/plugins/crawl/url_fuzzer.py).

It would be nice to get those url fuzzing tricks implemented into our url_fuzzer, giving credit in the source code to @mazen160 for his ideas on how to create the interesting files.

**Comment by andresriancho** [comment]
@alexandreborgo sure! Please go ahead!

I'm a little bit worried about the overlap between things that `bfac` does and other w3af plugins. For example, `find_dvcs.py` does search for `.gitignore`. So that one shouldn't be added to `url_fuzzer`.

`dot_ds_store.py` does find `.DS_Store` files, which are also mentioned in `bfac`.

Other extensions and transformations might have the same problem. I recommend doing an in-depth analysis of each path transformation mentioned here: https://github.com/mazen160/bfac/blob/master/bfac#L568-L726 and compare it with what w3af has.

For example, I found out that  there was an overlap with DSStore by running:
```
find . -type f -print0 | xargs -0 grep DS_Store
```

And inspecting the output.

Could you do that analysis and after that write the code? Please write the analysis results here in a comment.

**Comment by alexandreborgo** [comment]
Yes you're right, I hadn't thought about overlapping with other plugins.

Doing an analysis:
- bfac backup_testing_level1, backup_testing_level2 and backup_testing_level3: 
 	- change/add extension (backup, copy, save, etc), some are find by `url_fuzzer.py`, few aren't
 	- file transformation (inserting `Copy_of_` for example), no result on w3af
- bfac backup_testing_level4:
 	- files related to archive and compression, most are find by `url_fuzzer.py`, few aren't
 	- `bfac` looks for `site + file_path + '.'` and `site + file_path + '::$DATA'`
- bfac backup_testing_level5: 
	- files related to VCS, `find_dvcs.py` can find
	- files related to IDE setting `.idea/*`, seems like `pykto.py` looks for it
	- files related to composer, same as IDE setting

Note: some other plugins like `dir_file_bruter` and `pykto` try to find common files with these extensions, for examples `index.html.zip`, `.htaccess.bak`.

So `url_fuzzer` shouldn't look for common files, other plugins already take care of it.

What do you think about it ?

**Comment by andresriancho** [comment]
Thanks for the detailed and really good analysis!

I believe that `dir_file_bruter` is not that similar to what `bfac` or `url_fuzzer` do. `dir_file_bruter` will brute-force directories and files, without even caring about the known directories / files, while `bfac` and `url_fuzzer` do (mostly) transformations on existing URLs.

We shouldn't care much about overlap with `dir_file_bruter`.

So, to summarize:
 * [bfac level 5](https://github.com/mazen160/bfac/blob/master/bfac#L686-L726) should NOT be added to `url_fuzzer` because it is covered by other w3af plugins.
 * [bfac level 4](https://github.com/mazen160/bfac/blob/master/bfac#L663-L684) to *bfac level 1* we only need to add a few of the tests to `url_fuzzer`

Please go ahead and improve the plugin like that, thanks!

**Comment by andresriancho** [comment]
@alexandreborgo may I ask what MSSIS2018 is?

I received a few pull requests and comments from you guys in the last month(s). It seems to be a university class in France? If so, I'm open to talk with you guys about ways to improve our workflow / collaboration.

Thanks!
